python/lsst/daf/butler/cli/cliLog.py

```diff
@@ -109,48 +101,57 @@ def initLog(cls, longlog):
-            # configure lsst.log.
-            if not os.path.isfile(os.environ.get("LSST_LOG_CONFIG", "")):
-                lsstLog.configure_prop(_LOG_PROP.format(
-                    cls.lsstLog_longLogFmt if longlog else cls.lsstLog_normalLogFmt))
-            cls._recordComponentSetting(None)
-            pythonLogger = logging.getLogger()
-            pythonLogger.setLevel(logging.INFO)
-            cls._lsstLogHandler = lsstLog.LogHandler()
-            # Forward all Python logging to lsstLog
-            pythonLogger.addHandler(cls._lsstLogHandler)
+            # Ensure that log messages are forwarded back to python.
+            # Also allow MDC to be forwarded in log records if it has
+            # been set.
+            lsstLog.configure_pylog_MDC("DEBUG")
+
+            # Forward python lsst.log messages directly to python logging.
+            # This can bypass the C++ layer entirely but requires that
+            # MDC is set via ButlerMDC, rather than in lsst.log.
+            lsstLog.usePythonLogging()
+
+        if longlog:
+
+            # Want to create our own Formatter so that we can get high
+            # precision timestamps. This requires we attach our own
+            # default stream handler.
+            defaultHandler = logging.StreamHandler()
+            formatter = PrecisionLogFormatter(fmt=cls.pylog_longLogFmt, style="{")
+            defaultHandler.setFormatter(formatter)
+
+            logging.basicConfig(level=logging.INFO,
+                                force=True,
+                                handlers=[defaultHandler],
+                                )
+
         else:
-            cls._recordComponentSetting(None)
-            if longlog:
-                logging.basicConfig(level=logging.INFO,
-                                    format=cls.pylog_longLogFmt,
-                                    datefmt=cls.pylog_longLogDateFmt)
-            else:
-                logging.basicConfig(level=logging.INFO, format=cls.pylog_normalFmt)
+            logging.basicConfig(level=logging.INFO, format=cls.pylog_normalFmt, style="{")
+
+        # Initialize root logger level.
+        cls._setLogLevel(None, "INFO")
 
         # also capture warnings and send them to logging
         logging.captureWarnings(True)
 
-        # remember this call
-        cls.configState.append((cls.initLog, longlog))
+        # Create a record factory that ensures that an MDC is attached
+        # to the records. By default this is only used for long-log
+        # but always enable it for when someone adds a new handler
+        # that needs it.
+        old_factory = logging.getLogRecordFactory()
 
-    @classmethod
-    def getHandlerId(cls):
-        """Get the id of the lsst.log handler added to the python logger.
+        def record_factory(*args, **kwargs):
+            record = old_factory(*args, **kwargs)
+            record.MDC = ButlerMDC._MDC
+            return record
```

Comment:
`configure_pylog_MDC` already does the same so we are going to have two factories creating `MDC` objects which is not optimal performance-wise. Maybe we need additional flag to `configure_pylog_MDC` to make `_MDC` part optional?

Reply:
I was wondering about that. The lsst.log factory is called first and then the python one overrides it. I was pondering making this code merge the MDC from the lsst.log side but it's not really worth it given that no butler code will ever set the lsst.log MDC.  I'll add a small change to `configure_pylog_MDC` to make it skip the MDC part if the MDC class is None.

Reply:
See lsst/log#57